File: src/status.ts
export type BranchStatus = {
  raw: string;
  name: string;
  upstream?: string | undefined;
  ahead: number;
  behind: number;
  gone: boolean;
  metadata?: string | undefined;
};

export type StatusCategory = "modified" | "staged" | "untracked" | "deleted" | "renamed" | "conflict" | "other";

export type StatusItem = {
  raw: string;
  code: string;
  path: string;
  category: StatusCategory;
};

export type ParsedStatus = {
  raw: string;
  branch?: BranchStatus | undefined;
  items: StatusItem[];
  changed: boolean;
};

const CONFLICT_CODES = new Set(["DD", "AU", "UD", "UA", "DU", "AA", "UU"]);

export function hasChangesOrDivergence(status: string): boolean {
  for (const line of status.split(/\r?\n/)) {
    if (line.startsWith("## ")) {
      if (line.includes("[") && line.includes("]")) {
        return true;
      }
      continue;
    }

    if (line !== "") {
      return true;
    }
  }

  return false;
}

export function canFastForward(status: ParsedStatus): boolean {
  const branch = status.branch;
  if (branch === undefined || branch.name === "") {
    return false;
  }

  return (
    branch.upstream !== undefined &&
    !branch.gone &&
    branch.behind > 0 &&
    branch.ahead === 0 &&
    status.items.length === 0
  );
}

export function parseStatus(status: string): ParsedStatus {
  const lines = status.split(/\r?\n/).filter((line) => line !== "");
  const branchLine = lines.find((line) => line.startsWith("## "));
  const branch = branchLine === undefined ? undefined : parseBranchLine(branchLine);
  const items = lines.filter((line) => !line.startsWith("## ")).map(parseStatusLine);

  return {
    raw: status,
    branch,
    items,
    changed: hasChangesOrDivergence(status),
  };
}

export function parseBranchLine(line: string): BranchStatus {
  const raw = line;
  const content = line.startsWith("## ") ? line.slice(3) : line;
  const metadataMatch = content.match(/\s\[(.+)]$/);
  const metadata = metadataMatch?.[1];
  const branchPart = metadataMatch === null ? content : content.slice(0, metadataMatch.index).trimEnd();
  const upstreamSeparator = branchPart.indexOf("...");
  const name = upstreamSeparator === -1 ? branchPart : branchPart.slice(0, upstreamSeparator);
  const upstream = upstreamSeparator === -1 ? undefined : branchPart.slice(upstreamSeparator + 3);

  return {
    raw,
    name,
    upstream: upstream === "" ? undefined : upstream,
    ahead: parseMetadataCount(metadata, "ahead"),
    behind: parseMetadataCount(metadata, "behind"),
    gone: metadata?.includes("gone") ?? false,
    metadata,
  };
}

export function parseStatusLine(line: string): StatusItem {
  const code = line.slice(0, 2);
  const filePath = line.length > 3 ? line.slice(3) : "";

  return {
    raw: line,
    code,
    path: filePath,
    category: categorizeStatus(code, filePath),
  };
}

function categorizeStatus(code: string, filePath: string): StatusCategory {
  if (CONFLICT_CODES.has(code) || code.includes("U")) {
    return "conflict";
  }

  if (code === "??") {
    return "untracked";
  }

  if (code.includes("R") || filePath.includes(" -> ")) {
    return "renamed";
  }

  if (code.includes("D")) {
    return "deleted";
  }

  const indexStatus = code[0] ?? " ";
  const worktreeStatus = code[1] ?? " ";

  if (indexStatus !== " " && indexStatus !== "?") {
    return "staged";
  }

  if (worktreeStatus !== " ") {
    return "modified";
  }

  return "other";
}

function parseMetadataCount(metadata: string | undefined, key: "ahead" | "behind"): number {
  if (metadata === undefined) {
    return 0;
  }

  const match = metadata.match(new RegExp(`${key} (\\d+)`));
  if (match?.[1] === undefined) {
    return 0;
  }

  return Number.parseInt(match[1], 10);
}

File: internal/status/status.go
package status

import (
	"regexp"
	"strconv"
	"strings"
)

type BranchStatus struct {
	Raw      string
	Name     string
	Upstream string
	Ahead    int
	Behind   int
	Gone     bool
	Metadata string
}

type Category string

const (
	Modified  Category = "modified"
	Staged    Category = "staged"
	Untracked Category = "untracked"
	Deleted   Category = "deleted"
	Renamed   Category = "renamed"
	Conflict  Category = "conflict"
	Other     Category = "other"
)

type Item struct {
	Raw      string
	Code     string
	Path     string
	Category Category
}

type Parsed struct {
	Raw     string
	Branch  *BranchStatus
	Items   []Item
	Changed bool
}

var (
	metadataPattern = regexp.MustCompile(`\s\[(.+)]$`)
	countPatterns   = map[string]*regexp.Regexp{
		"ahead":  regexp.MustCompile(`ahead (\d+)`),
		"behind": regexp.MustCompile(`behind (\d+)`),
	}
	conflictCodes = map[string]struct{}{
		"DD": {},
		"AU": {},
		"UD": {},
		"UA": {},
		"DU": {},
		"AA": {},
		"UU": {},
	}
)

func HasChangesOrDivergence(raw string) bool {
	for _, line := range splitLines(raw) {
		if strings.HasPrefix(line, "## ") {
			if strings.Contains(line, "[") && strings.Contains(line, "]") {
				return true
			}
			continue
		}
		if line != "" {
			return true
		}
	}
	return false
}

func CanFastForward(parsed Parsed) bool {
	if parsed.Branch == nil || parsed.Branch.Name == "" {
		return false
	}
	return parsed.Branch.Upstream != "" &&
		!parsed.Branch.Gone &&
		parsed.Branch.Behind > 0 &&
		parsed.Branch.Ahead == 0 &&
		len(parsed.Items) == 0
}

func Parse(raw string) Parsed {
	lines := []string{}
	for _, line := range splitLines(raw) {
		if line != "" {
			lines = append(lines, line)
		}
	}

	var branch *BranchStatus
	items := []Item{}
	for _, line := range lines {
		if strings.HasPrefix(line, "## ") {
			parsedBranch := ParseBranchLine(line)
			branch = &parsedBranch
			continue
		}
		items = append(items, ParseStatusLine(line))
	}

	return Parsed{
		Raw:     raw,
		Branch:  branch,
		Items:   items,
		Changed: HasChangesOrDivergence(raw),
	}
}

func ParseBranchLine(line string) BranchStatus {
	raw := line
	content := line
	if strings.HasPrefix(content, "## ") {
		content = strings.TrimPrefix(content, "## ")
	}

	metadata := ""
	branchPart := content
	if match := metadataPattern.FindStringSubmatchIndex(content); match != nil {
		metadata = content[match[2]:match[3]]
		branchPart = strings.TrimRight(content[:match[0]], " ")
	}

	name := branchPart
	upstream := ""
	if index := strings.Index(branchPart, "..."); index != -1 {
		name = branchPart[:index]
		upstream = branchPart[index+3:]
	}

	return BranchStatus{
		Raw:      raw,
		Name:     name,
		Upstream: upstream,
		Ahead:    parseMetadataCount(metadata, "ahead"),
		Behind:   parseMetadataCount(metadata, "behind"),
		Gone:     strings.Contains(metadata, "gone"),
		Metadata: metadata,
	}
}

func ParseStatusLine(line string) Item {
	code := line
	if len(code) > 2 {
		code = line[:2]
	}
	filePath := ""
	if len(line) > 3 {
		filePath = line[3:]
	}

	return Item{
		Raw:      line,
		Code:     code,
		Path:     filePath,
		Category: categorize(code, filePath),
	}
}

func categorize(code string, filePath string) Category {
	if _, ok := conflictCodes[code]; ok || strings.Contains(code, "U") {
		return Conflict
	}
	if code == "??" {
		return Untracked
	}
	if strings.Contains(code, "R") || strings.Contains(filePath, " -> ") {
		return Renamed
	}
	if strings.Contains(code, "D") {
		return Deleted
	}

	indexStatus := byte(' ')
	worktreeStatus := byte(' ')
	if len(code) > 0 {
		indexStatus = code[0]
	}
	if len(code) > 1 {
		worktreeStatus = code[1]
	}

	if indexStatus != ' ' && indexStatus != '?' {
		return Staged
	}
	if worktreeStatus != ' ' {
		return Modified
	}
	return Other
}

func parseMetadataCount(metadata string, key string) int {
	if metadata == "" {
		return 0
	}
	match := countPatterns[key].FindStringSubmatch(metadata)
	if len(match) < 2 {
		return 0
	}
	parsed, err := strconv.Atoi(match[1])
	if err != nil {
		return 0
	}
	return parsed
}

func splitLines(raw string) []string {
	return strings.Split(strings.ReplaceAll(raw, "\r\n", "\n"), "\n")
}
